core/src/sumeragi/view_change.rs

```diff
@@ -28,103 +30,153 @@ struct ProofPayload {
 }
 
 /// The proof of a view change. It needs to be signed by f+1 peers for proof to be valid and view change to happen.
-#[derive(Debug, Clone, Decode, Encode)]
+#[derive(Debug, Clone, Encode)]
 pub struct SignedProof {
-    signatures: SignaturesOf<ProofPayload>,
-    /// Collection of signatures from the different peers.
+    signatures: Vec<ViewChangeProofSignature>,
     payload: ProofPayload,
 }
 
+mod candidate {
+    use indexmap::IndexSet;
+    use parity_scale_codec::Input;
+
+    use super::*;
+
+    #[derive(Decode)]
+    struct SignedProofCandidate {
+        signatures: Vec<ViewChangeProofSignature>,
+        payload: ProofPayload,
+    }
+
+    impl SignedProofCandidate {
+        fn validate(self) -> Result<SignedProof, &'static str> {
+            self.validate_signatures()?;
+
+            Ok(SignedProof {
+                payload: self.payload,
+                signatures: self.signatures,
+            })
+        }
+
+        fn validate_signatures(&self) -> Result<(), &'static str> {
+            self.signatures
+                .iter()
+                .try_fold(IndexSet::new(), |mut acc, elem| {
+                    if !acc.insert(elem) {
+                        return Err("Duplicate signature in proof");
+                    }
```

Comment:
I'm not sure whether proofs are duplicated if peer is repeatedly requesting view change?

@Erigara @SamHSmith

Reply:
Since you've changed type from set to vector singnatures now can duplicate. 

You can look at `merge_signatures` function.